faststreams/input_stream.nim

```diff
@@ -1,192 +1,298 @@
+    stream.bufferStart = head
+
+  InputStreamHandle(s: stream)
+
+proc memoryInput*(mem: openarray[byte]): InputStreamHandle =
+  let head = unsafeAddr mem[0]
+  InputStreamHandle(s: InputStream(
+    head: head,
+    bufferEnd: offset(head, mem.len),
+    bufferEndPos: mem.len))
+
+proc memoryInput*(str: string): InputStreamHandle =
+  memoryInput str.toOpenArrayByte(0, str.len - 1)
+
+# TODO: Is this used, should we deprecate it?
+proc endPos*(s: InputStream): int =
+  doAssert s.vtable == nil or s.vtable.getLenSync != nil
   return s.bufferEndPos
 
-proc syncRead(s: var ByteStream): bool =
-  let bytesRead = s.reader(s.bufferStart, s.bufferSize)
+# TODO The return type here could be Option[Natural] if Nim had
+# the Option[range] optimisation that will make it equvalent to `int`.
+proc len*(s: InputStream): int {.raises: [Defect, IOError].} =
+  if s.vtable == nil:
+    distance(s.head, s.bufferEnd)
+  elif s.vtable.getLenSync != nil:
+    s.vtable.getLenSync(s)
+  else:
+    lengthUnknown
+
+template len*(s: AsyncInputStream): int =
+  len InputStream(s)
+
+proc bufferMoreDataSync(s: InputStream): bool =
+  # Returns true if more data was successfully buffered
+  if s.vtable == nil or s.vtable.readSync == nil:
+    return false
+
+  let bytesRead = s.vtable.readSync(s, s.bufferStart, s.bufferSize)
   if bytesRead == 0:
-    s.reader = nil
-    return true
+    # TODO close the input device
+    s.vtable = nil
+    return false
   else:
     s.bufferEnd = offset(s.bufferStart, bytesRead)
     s.bufferEndPos += bytesRead
-    return false
+    return true
 
-proc ensureBytes*(s: var ByteStream, n: int): bool =
+proc bufferMoreDataAsync(s: AsyncInputStream): Future[bool] {.async.} =
+  # Returns true if more data was successfully buffered
+  return false
+
+proc readable*(s: InputStream): bool =
+  if s.head != s.bufferEnd:
+    true
+  else:
+    s.bufferMoreDataSync()
+
+template readable*(sp: AsyncInputStream): bool =
+  let s = sp
+  if s.head != s.bufferEnd:
+    true
+  else:
+    faststreamsAwait bufferMoreDataAsync(s)
+
+proc readable*(s: InputStream, n: int): bool =
   if distance(s.head, s.bufferEnd) >= n:
     return true
 
-  if s.reader == nil:
+  if s.vtable == nil or s.vtable.readSync == nil:
     return false
 
+  # TODO
   doAssert false, "Multi-buffer reading will be implemented later"
 
-proc eof*(s: var ByteStream): bool =
-  if s.head != s.bufferEnd:
-    return false
+template readable*(sp: AsyncInputStream, n: int): bool =
+  let s = sp
 
-  if s.reader == nil:
+  if distance(s.head, s.bufferEnd) >= n:
     return true
 
-  return s.syncRead()
+  if s.vtable == nil:
+    return false
+
+  # TODO
+  doAssert false, "Multi-buffer reading will be implemented later"
 
-proc eob*(s: ByteStream): bool {.inline.} =
-  s.head != s.bufferEnd
+template close*(s: AsyncInputStream) =
+  close InputStream(s)
 
-proc peek*(s: ByteStream): byte {.inline.} =
+proc peek*(s: InputStream): byte {.inline.} =
```

Comment:
`peek` is rarely needed in practice, but you need it sometimes. In a `LL(1)` parser for example, you would often peek at the next character before you decide which stream consuming function to invoke next.

Reply:
Few more thoughts to share:

When you think about these APIs in isolation, it's easy to invoke the usual common wisdom that pre-conditions lead to danger and total functions are safer, but you if you look closer at actual usages, you'll find a common idiomatic pattern that emerges. Most of the code looks roughly like this:

```nim
while stream.readable:
  ...
  stream.peek
  ...
  stream.read
```

This is both quite natural to write and safe within the API.

Reply:
well:

```
while (let byte = stream.peek; byte.isSome):
  stream.advance...
```

safe and sound, no possibility for errors - arguably idiomatic emergent usage.

Reply:
Well, you must really dislike pre-conditions to suggest this. If every single user of the API has to write a weird loop like this, I'd be shy to recommend the library to anyone - it would be hard to imagine it being accepted in Nim's standard library in the future, which I consider valuable to us for interoperability purposes.

Furthermore, this idea that all functions should produce an optional copy of the data extracted from the stream really flies in the face of the "consume multiple bytes" use case. In this scenario, you want the data to sit in the stream buffers - you write your parser/procesesor/etc in a way that it consumes the data directly from these buffers.

It's easy to learn the idioms and the compiler will be able to enforce them in the future with `drnim`. If you still get the code wrong, you'll hit the asserts pretty fast. They are not triggered only in some corner cases.

Reply:
One exception to what I mentioned above about consuming multiple bytes is another API that I'll add soon. It will be called `readInto` and it will have the following signature:

```nim
proc readInto(stream: InputStream, destination: var openarray[byte]): bool
  # Returns `false` if EOF was reached before populating the destination buffer
```

It will interact with the buffer management scheme used in the stream to insert the destination buffer as a "page" within the buffering mechanism, so you get "zero copy" consumption. But such an API is useful only when you truly want to extract raw bytes from the stream and you are not interested in processing or transforming them in any way.

Reply:
> If every single user of the API has to write a weird loop like this,

well, the point kind of is to demonstrate that a safe api is possible - if it turns out to be useful, the syntax can be streamlined. 

drnim is a :unicorn: 

`var openarray` looks like a much more pragmatic way forward - perhaps though it should return an `int` with the number of bytes it managed to get out of there - and maybe it should be `peekInto` so that you can advance "custom" amounts of bugs.

Reply:
`drnim` was already shipped in Nim 1.2 and it will see significant development in the coming months. The pre-conditions present here are some of the most simple ones to enforce.

In general, I'm well familiar with the arguments against exceptions that you often make. These ideas are indeed shared by many. But you also often bake in there an argument against pre-conditions and I recognize this as a fringe opinion. I've never seen someone so vigorously attacking the ideas of design by contract. Can you point me to some specific criticisms here:

https://wiki.c2.com/?DesignByContract
https://en.wikipedia.org/wiki/Design_by_contract 

The most significant problem with the APIs that you suggest is that they lead to unnecessary memory allocations and copying of data. The APIs that exist now are easy enough to use correctly and the extra performance gains justify the presence of pre-conditions.

Reply:
> Returns `false` if EOF was reached before populating the destination buffer

well, I'm just pointing out that you need to solve partial reads before the EOF - what's correct depends on the use case.

contracts are fun toys, but at runtime they kind of suck and at compile time it's often too slow to be of practical use except for the most trivial of cases - ie there's a reason why nobody uses them, really even though they've been around for years. if I can get a similar effect with simpler tools - say type and function signatures, that seems... well.. just simpler :)